### PR TITLE
Feature 85 chat history thread

### PR DESCRIPTION
**Pull request by gyfantidis**
FIX 85
FIX 89

# FIX Tickets
 - #85: As a User I want to have chat history when discussing with a Gendox Agent
 - #89: As a DevOps Developer, I want to rename the Discord's '/ask' command to '/search' command.
 
**Description**:

- V20231027_094000__Add_thread_history_column.sql
  - Create columns if they don't exist

- CommonCommandUtility
  - create executeCommandCode method for all commands
  
- Create Message package for commands
  - ChatGendoxMessage
  - SearchGendoxMessage
  - ReplyGendoxMessage
  - ListenerService
  
- Create threadId metadata into CompletionMessageDTO
  - CompletionMessageDTO
  - EmbeddingsController
  
- Implement the user's referenced message
  - Listener
  
- Create if control for PostMapping Controllers (id = null)
  - DocumentController
  - OrganizationController
  - ProjectController
  - UserController
  
- Add projectId & threadId
  - Message
  - MessageDTO
  - MessageConverter

- Create getByProjectId method
  - ProjectAgentService
  
- Create Reply gendox command
  - ReplyGendox
  
- Rename ask gendox to search gendox
  - SearchGendox

### Related Ticket: #85

### CODE REVIEW

File: gendox-core-api/src/main/java/dev/ctrlspace/gendox/gendoxcoreapi/discord/commands/SearchGendox.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import java.util.ArrayList;
+import java.util.List;
+
+
+@Component
+public class SearchGendox implements ICommand {
```
[the remark this answers is not in the document]
Comment (gyfantidis) [reply gendox-core-api/src/main/java/dev/ctrlspace/gendox/gendoxcoreapi/discord/commands/SearchGendox.java:19]: This is the AskGendox command, we renamed it.

File: gendox-core-api/src/main/java/dev/ctrlspace/gendox/gendoxcoreapi/discord/commands/ReplyGendox.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+import org.slf4j.Logger;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+
+import java.util.ArrayList;
+import java.util.List;
+
+@Component
+public class ReplyGendox implements ICommand {
```
[the remark this answers is not in the document]
Comment (gyfantidis) [reply gendox-core-api/src/main/java/dev/ctrlspace/gendox/gendoxcoreapi/discord/commands/ReplyGendox.java:20]: hmm i don't know let's say we will have 2 options for reply

File: gendox-core-api/src/main/java/dev/ctrlspace/gendox/gendoxcoreapi/discord/utils/CommonCommandUtility.java [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+
+
+            // Get the message content from the event
+            String question = getTheQuestion(event);
+            channel.sendMessage(authorName + ", thank you for the question: \n- " + question + "\n\uD83E\uDD16 Thinking... \uD83E\uDD16").queue();
+
+            if (command.equals(DiscordGendoxConstants.CHAT_GENDOX)) {
+                CompletionMessageDTO completionMessageDTO = listenerService.completionForQuestion(question, channelName, jwtToken, threadId);
+                chatGendoxMessage.chatMessage(channel, completionMessageDTO);
+                replyButton(channel);
```
[the remark this answers is not in the document]
Comment (gyfantidis) [reply gendox-core-api/src/main/java/dev/ctrlspace/gendox/gendoxcoreapi/discord/utils/CommonCommandUtility.java:89]: may be until if's I use somethink like this ?
```
switch (command) {
    case DiscordGendoxConstants.CHAT_GENDOX:
        CompletionMessageDTO completionMessageDTO = listenerService.completionForQuestion(question, channelName, jwtToken, threadId);
        chatGendoxMessage.chatMessage(channel, completionMessageDTO);
        replyButton(channel);
        break;
    case DiscordGendoxConstants.SEARCH_GENDOX:
        List<DocumentInstanceSection> documentInstanceSections = listenerService.semanticSearchForQuestion(question, channelName, jwtToken, threadId);
        searchGendoxMessage.searchMessage(channel, documentInstanceSections, projectId);
        break;
    case DiscordGendoxConstants.REPLY_GENDOX:
        CompletionMessageDTO completionMessageDTO = listenerService.completionForQuestion(question, channelName, jwtToken, threadId);
        chatGendoxMessage.chatMessage(channel, completionMessageDTO);
        break;
    default:
        throw new GendoxException("COMMAND_NOT_EXIST", "This command does not exist", HttpStatus.BAD_REQUEST);
}
```

File: gendox-core-api/src/main/java/dev/ctrlspace/gendox/gendoxcoreapi/utils/SecurityUtils.java [hunk 4]
```diff
@@ -135,9 +142,13 @@ public boolean hasAuthorityToAllRequestedProjectId() {
     }
 
     public UUID getUserId() {
+        try {
             Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
             JwtDTO jwtDTO = jwtUtils.toJwtDTO((Jwt) authentication.getPrincipal());
             return UUID.fromString(jwtDTO.getUserId());
+        } catch (Exception e){
+            return null;
```
[the remark this answers is not in the document]
Comment (gyfantidis) [reply gendox-core-api/src/main/java/dev/ctrlspace/gendox/gendoxcoreapi/utils/SecurityUtils.java:150]:
```
public UUID getUserId() {
        try {
            Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
            JwtDTO jwtDTO = jwtUtils.toJwtDTO((Jwt) authentication.getPrincipal());
            return UUID.fromString(jwtDTO.getUserId());
        } catch (Exception e){
            logger.warn("An exception occurred while trying to get the user ID: " + e.getMessage());
            return null;
        }
    }
    ```

File: gendox-core-api/src/main/java/dev/ctrlspace/gendox/gendoxcoreapi/services/TypeService.java [hunk 5]
```diff
@@ -61,6 +61,12 @@ public Type getDocumentSplitterTypeByName(String splitterName){
 
     }
 
+    public Type getUserRoleTypeByUserType(String userType){
```
[the remark this answers is not in the document]
Comment (gyfantidis) [reply gendox-core-api/src/main/java/dev/ctrlspace/gendox/gendoxcoreapi/services/TypeService.java:64]: I deleted it, there was the same method in the same class

```
public Type getUserTypeByName(String userTypeName){
        return typeRepository.findByTypeCategoryAndName("USER_TYPE", userTypeName)
                .orElseThrow(() -> new GendoxRuntimeException(HttpStatus.NOT_FOUND, "USER_TYPE_NOT_FOUND", "User field type not found with name: " + userTypeName));

    }   
 ```

**Review by sekasx — APPROVED** [review]

File: gendox-core-api/src/main/java/dev/ctrlspace/gendox/gendoxcoreapi/discord/utils/CommonCommandUtility.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (sekasx) [reply gendox-core-api/src/main/java/dev/ctrlspace/gendox/gendoxcoreapi/discord/utils/CommonCommandUtility.java:89]: Leave it as it is, we will find a better way without any `if` or `case`

**Pull request merged**